### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Dzemlianoi**

### CODE REVIEW

File: app/controllers/order_steps_controller.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    step_to(:complete) and return if last_order.confirmed? && !on_step?(:complete)
+    check_info_steps if on_step? :confirm
+    step_to(:confirm) and return if last_order.in_confirmation? && !on_step?(:confirm)
+    step_to(:confirm) if on_step?(:complete) && !last_order.confirmed?
+  end
+
+  def update
+    byebug
+    @updating_result = @form.update(step, order_params)
+    step_to next_step and return if @updating_result.eql?(true)
```
Comment (andrii-novikov) [inline app/controllers/order_steps_controller.rb]: if result == true ??

File: app/controllers/users/registrations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        set_flash_message! :notice, :"signed_up_but_#{resource.inactive_message}"
+        expire_data_after_sign_in!
+        respond_with resource, location: after_inactive_sign_up_path_for(resource)
+      end
+    else
+      clean_up_passwords resource
+      set_minimum_password_length
+      respond_with resource
+    end
+  end
```
Comment (yamnich) [inline app/controllers/users/registrations_controller.rb]: Too long method. It is better to divide it and avoid using so many ifs.

Devise Controller. Refactored 2 else blocks into "super"

File: app/controllers/order_steps_controller.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  include Wicked::Wizard
+
+  before_action :go_root_unless_order
+  before_action :initialize_form
+
+  steps :fast_sign, :address, :delivery, :payment, :confirm, :complete
+
+  def show
+    case step
+      when :fast_sign then next_step and return if current_user
```
Comment (yamnich) [inline app/controllers/order_steps_controller.rb]: Do you use "and" for some purpose? Is it wicked requirements?

"and" is working for redirects while "&&" don't

File: app/forms/order_steps_form.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+  def update(step, params)
+    case step
+      when :address
+        @order.update_attributes(use_billing: 0) unless params.has_key? :use_billing
+        if params.has_key? :use_billing
+          params[:shipping_address] = params[:billing_address]
+          @order.update_attribute(:use_billing, 1)
+        end
+        !!create_addresses(params) & orders_saved?
```
Comment (yamnich) [inline app/forms/order_steps_form.rb]: it is better to move things like this to method with meaningful name

Renamed and moved to another method. Done

File: app/forms/order_steps_form.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    return @billing_address.update(params) if @order.billing_address
+    @billing_address = @order.addresses.billing.create(params)
+    @billing_address.persisted?
+  end
+
+  def create_shipping(params)
+    params = params.merge(kind: :shipping)
+    return @shipping_address.update(params) if @order.shipping_address
+    @shipping_address = @order.addresses.shipping.create(params)
+    @billing_address.persisted?
```
Comment (yamnich) [inline app/forms/order_steps_form.rb]: 1. are you sure that @billing_address should be here?
2. these two methods are almost the same, how can we avoid duplication?

move to metaprogramming. Done!

File: app/forms/order_steps_form.rb [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    return unless Delivery.find_by_id(delivery_id)
+    @order.update(delivery_id: delivery_id)
+  end
+
+  def create_addresses(params)
+    create_billing(params[:billing_address]) && create_shipping(params[:shipping_address])
+  end
+
+  def orders_saved?
+    @order.addresses.shipping.present? && @order.addresses.billing.present?
```
Comment (yamnich) [inline app/forms/order_steps_form.rb]: in such cases
addresses = @order.addresses
addresses.shipping.any? ...

done

File: app/models/address.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    country.translations[I18n.locale.to_s] || country.name
+  end
+
+  def add_username
+    if self.addressable.instance_of? User
+      self.first_name = self.addressable.first_name if self.first_name.nil?
+      self.last_name = self.addressable.last_name if self.last_name.nil?
+    else
+      self.first_name = self.addressable.user.first_name if self.first_name.nil?
+      self.last_name = self.addressable.user.last_name if self.last_name.nil?
```
Comment (yamnich) [inline app/models/address.rb]: some_addressable = self.addressable.instance_of? User ? self.addressable : self.addressable.user
 self.first_name = some_addressable.first_name if self.first_name.nil?
 self.last_name = some_addressable.last_name if self.last_name.nil?

Done!

File: app/models/user.rb [hunk 8]
```diff
@@ -1,2 +1,97 @@
...
+
+  def shipping_address
+    addresses.find_by(kind: :shipping)
+  end
+
+  def self.from_omniauth(auth)
+    @auth = auth
+    case
+      when where(email: @auth.info.email).without_provider.present?
+        @user = where(email: @auth.info.email).without_provider.first.update_attributes(
```
Comment (yamnich) [inline app/models/user.rb]: 1.  where(email: @auth.info.email).without_provider used twice. You can move it to scope

Done

File: app/models/user.rb [hunk 9]
```diff
@@ -1,2 +1,97 @@
...
+  def email_required?
+    super && provider.blank? && guest_token.blank?
+  end
+
+  def password_required?
+    super && provider.blank? && guest_token.blank?
+  end
+
+  def confirmation_required?
+    super && provider.blank? && guest_token.blank?
```
Comment (yamnich) [inline app/models/user.rb]: 3 equal methods. why?

move to define_method. Done!

File: app/views/orders/show.html.haml [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    .col-sm-3
+      %h3.general-subtitle.mt-0= t('users.addresses.billing_address')
+      %p.general-address
+        #{@order.billing_address.first_name} #{@order.billing_address.last_name}
+        %br
+        #{@order.billing_address.address}
+        %br
+        #{@order.billing_address.city} #{@order.billing_address.zip}
+        %br
+        #{@order.billing_address.country_name}
```
Comment (yamnich) [inline app/views/orders/show.html.haml]: can we move it to partial?

moved to partials. done

**Comment by yamnich** [comment]
When we started book store I have told you that form objects should be used, but I don't see any. Why?

File: app/controllers/order_steps_controller.rb [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    when :address, :delivery, :payment
+      step_to :fast_sign and return unless current_user
+      step_to :complete and return if last_active_order.in_confirmation?
+      step_to :confirm and return if last_active_order.filled? && !order_params[:editing].present?
+    when :confirm then return check_info_steps
+    when :complete
+      step_to :fast_sign and return unless current_user
+      step_to :confirm and return unless last_active_order.in_confirmation?
+    end
+    render_wizard
```
Comment (yamnich) [inline app/controllers/order_steps_controller.rb]: It is really complex method for controller. too many if. Can we move it somewhere and divide?

File: app/helpers/devise_helper.rb [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    return '' if resource.errors.empty?
+    messages = resource.errors.full_messages.map { |msg| content_tag(:li, msg) }.join
+    html = <<-HTML
+    <div class="text-center alert alert-error alert-danger">
+       <button type="button" class="close" data-dismiss="alert">×</button>
+       #{messages}
+    </div>
+    HTML
+    html.html_safe
+  end
```
Comment (yamnich) [inline app/helpers/devise_helper.rb]: the same as in application helper, can we use parent class or module here?

File: app/mailers/order_mailer.rb [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    @user = user
+    mail(to: @user.email,
+         subject: 'Confirm your purchase',
+         template_path: 'orders/mailers',
+         template_name: 'confirm')
+  end
+
+  def treating_send(user, order)
+    @order = order
+    @user = user
```
Comment (yamnich) [inline app/mailers/order_mailer.rb]: @order = order
 @user = user

this is used in many methods, can we move it somewhere?

File: app/models/order.rb [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+    ORDERING[DEFAULT_SORT_KEY]
+  end
+
+  def billing_address
+    addresses.find_by(kind: 'billing')
+  end
+
+  def shipping_address
+    addresses.find_by(kind: 'shipping')
+  end
```
Comment (yamnich) [inline app/models/order.rb]: metaprogramming?

yeap! Done

File: app/models/order.rb [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    !order_items.find_by(book: book_id).nil?
+  end
+
+  def subtotal_more_than_discount?(coupon)
+    subtotal_price > coupon.discount
+  end
+
+  def send_confirmation
+    update_attributes(confirmation_token: Devise.friendly_token)
+    OrderMailer.confirmation_send(user, self).deliver_now
```
Comment (yamnich) [inline app/models/order.rb:96]: it is not very good to send emails from model, you can use observers

File: app/models/order.rb [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+  def treat_proccessing
+    self.confirmation_token = nil
+    OrderMailer.treating_send(user, self).deliver_now
+  end
+
+  def send_success
+    OrderMailer.success_letter(user, self).deliver_later
+  end
+
+  def recalculate_total(*)
```
Comment (yamnich) [inline app/models/order.rb]: why do you use (*)?

now use (_). Done

File: app/models/order.rb [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+
+  def subtotal_price
+    order_items.map { |item| item.book[:price] * item.quantity }.inject(&:+)
+  end
+
+  def total_price
+    return 0.00 if order_items.empty?
+    price = subtotal_price + delivery_price - discount
+    price.positive? ? price : 0.00
+  end
```
Comment (yamnich) [inline app/models/order.rb]: it can be better to move all prices calculations to some service

File: app/forms/order_steps_form.rb [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  def create_delivery(delivery_id)
+    return unless Delivery.find_by(id: delivery_id)
+    @order.update_attribute(:delivery_id, delivery_id)
+  end
+
+  def both_addresses_present?
+    addresses = @order.addresses
+    addresses.shipping.present? & addresses.billing.present?
```
Comment (yamnich) [inline app/forms/order_steps_form.rb:71]: && ?

File: app/models/order.rb (hunk 15 above)
Comment (yamnich) [reply app/models/order.rb:96]: When you use update_attributes save is calling so you can use before/after save callback with observer and check if confirmation_token was changed. It is just possible variant :)

**Review by yamnich — APPROVED** [review]

**Pull request merged**